### PR TITLE
Add information about PodLifecycleSleepAction (KEP3960)

### PR DESCRIPTION
**Pull request by AxeZhan**
PodLifecycleSleepAction updates as part of https://github.com/kubernetes/enhancements/issues/3960

### CODE REVIEW

**Comment by AxeZhan** [comment]
/remove-language de es fr pl pt zh
/remove-sig release

**Comment by AxeZhan** [comment]
/cc @thockin 
Not sure whether if this is enough😢. And if not what should be added, and to where🤦.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -169,6 +169,7 @@ For a reference to old feature gates that are removed, please refer to
 | `PodDisruptionConditions` | `false` | Alpha | 1.25 | 1.25 |
 | `PodDisruptionConditions` | `true` | Beta | 1.26 | |
 | `PodHostIPs` | `false` | Alpha | 1.28 | |
+| `PodLifecycleSleepAction` | `false` | Alpha | 1.29 | |
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:172]: This should appear of the `PodIndexLabel`

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 2]
```diff
@@ -663,6 +664,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `PodDisruptionConditions`: Enables support for appending a dedicated pod condition indicating that the pod is being deleted due to a disruption.
 - `PodHostIPs`: Enable the `status.hostIPs` field for pods and the {{< glossary_tooltip term_id="downward-api" text="downward API" >}}.
   The field lets you expose host IP addresses to workloads.
+- `PodLifecycleSleepAction`: Set the `sleep.seconds` filed in Container Lifecycle Hooks to let containers pause for a specified duration.
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:667]: 'filed' -> 'field'.

Please clarify if the gate enables this setting or the gate will set it automatically.

File: content/en/docs/reference/kubernetes-api/workload-resources/pod-v1.md [hunk 3]
```diff
@@ -1592,7 +1592,16 @@ LifecycleHandler defines a specific action that should be taken in a lifecycle h
 
     Optional: Host name to connect to, defaults to the pod IP.
 
+- **sleep** (SleepAction)
```
Comment (tengqm) [inline content/en/docs/reference/kubernetes-api/workload-resources/pod-v1.md:1595]: Don't touch this file yet.
This file is auto-generated and will be refreshed once 1.29 is out of the door.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 2 above)
Comment (AxeZhan) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md]: Updated, how about now?

File: content/en/docs/reference/kubernetes-api/workload-resources/pod-v1.md (hunk 3 above)
Comment (AxeZhan) [reply content/en/docs/reference/kubernetes-api/workload-resources/pod-v1.md]: Okay, how is this file generated? Should I pay attention to it?

Comment (tengqm) [reply content/en/docs/reference/kubernetes-api/workload-resources/pod-v1.md]: No need to pay attention to it at the moment.
v1.29 website (including your PR) won't get published until 1.29 becomes a thing.

File: content/en/docs/concepts/containers/container-lifecycle-hooks.md [hunk 4]
```diff
@@ -55,12 +55,13 @@ There are two types of hook handlers that can be implemented for Containers:
 * Exec - Executes a specific command, such as `pre-stop.sh`, inside the cgroups and namespaces of the Container.
 Resources consumed by the command are counted against the Container.
 * HTTP - Executes an HTTP request against a specific endpoint on the Container.
+* Sleep - Executes a sleep command to pause a container for a specified duration. (Note: "Sleep" is available when the feature-gate PodLifecycleSleepAction is enabled)
```
Comment (tengqm) [inline content/en/docs/concepts/containers/container-lifecycle-hooks.md:58]: Wrap the long line, pls.
```suggestion
* Sleep - Executes a sleep command to pause a container for a specified duration.
  The "Sleep" action is available when the [feature gate](/docs/reference/command-line-tool-reference/feagure-gates/)
  `PodLifecycleSleepAction` is enabled.
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 5]
```diff
@@ -664,6 +665,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `PodHostIPs`: Enable the `status.hostIPs` field for pods and the {{< glossary_tooltip term_id="downward-api" text="downward API" >}}.
   The field lets you expose host IP addresses to workloads.
 - `PodIndexLabel`: Enables the Job controller and StatefulSet controller to add the pod index as a label when creating new pods. See [Job completion mode docs](/docs/concepts/workloads/controllers/job#completion-mode) and [StatefulSet pod index label docs](/docs/concepts/workloads/controllers/statefulset/#pod-index-label) for more details.
+- `PodLifecycleSleepAction`: Enable the sleepAction in Container Lifecycle Hooks.
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:668]: What is the word "sleepAction"? Is it a user-facing field name or a variable that only makes sense to developers?

Comment (AxeZhan) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:668]: It's not a field name, it's we use field name + "Action" to describe a action in `LifecycleHandler `.

Example:
```
- **exec** (ExecAction)

  Exec specifies the action to take.

  <a name="ExecAction"></a>
  *ExecAction describes a "run in container" action.*

  - **exec.command** ([]string)

    Command is the command line to execute inside the container, the working directory for the command  is root ('/') in the container's filesystem. The command is simply exec'd, it is not run inside a shell, so traditional shell instructions ('|', etc) won't work. To use a shell, you need to explicitly call out to that shell. Exit status of 0 is treated as live/healthy and non-zero is unhealthy.

- **httpGet** (HTTPGetAction)

  HTTPGet specifies the http request to perform.
```

Comment (tengqm) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:668]: ExecAction is a "definition" that can be find elsewhere in the API reference, just like "Metadata". A definition usually starts with a capital letter, so I'm confused what is a "sleepAction".

Comment (AxeZhan) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:668]: Sorry, my bad. It should also start with a capital letter.
SleepAction is also a "definition", currently you can find it in k/k repo:
```
// ExecAction describes a "run in container" action.
type ExecAction struct {
	// Command is the command line to execute inside the container, the working directory for the
	// command  is root ('/') in the container's filesystem.  The command is simply exec'd, it is
	// not run inside a shell, so traditional shell instructions ('|', etc) won't work.  To use
	// a shell, you need to explicitly call out to that shell.
	// +optional
	Command []string
}

// SleepAction describes a "sleep" action.
type SleepAction struct {
	// Seconds is the number of seconds to sleep.
	Seconds int64
}
```
However, I didn't find a correct place in this repo to put it.

**Comment by tengqm** [comment]
Please squash the PR into a single commit.
The definitions (if appears in the `types.go` file) will be refreshed by the reference generator. Don't worry about it at the moment.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 6]
```diff
@@ -664,6 +665,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `PodHostIPs`: Enable the `status.hostIPs` field for pods and the {{< glossary_tooltip term_id="downward-api" text="downward API" >}}.
   The field lets you expose host IP addresses to workloads.
 - `PodIndexLabel`: Enables the Job controller and StatefulSet controller to add the pod index as a label when creating new pods. See [Job completion mode docs](/docs/concepts/workloads/controllers/job#completion-mode) and [StatefulSet pod index label docs](/docs/concepts/workloads/controllers/statefulset/#pod-index-label) for more details.
+- `PodLifecycleSleepAction`: Enable the SleepAction in Container Lifecycle Hooks.
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:668]:
```suggestion
- `PodLifecycleSleepAction`: Enables the `sleep` action in Container lifecycle hooks.
```

**Comment by tengqm** [comment]
/approve
/hold

The hold can be lifted when the upstream implementation is merged.

**Comment by AxeZhan** [comment]
> The hold can be lifted when the upstream implementation is merged.

What does `the upstream implementation` refer to? If it's code changes in k/k repo, then we had it merged before. https://github.com/kubernetes/kubernetes/pull/119026

File: content/en/docs/concepts/containers/container-lifecycle-hooks.md [hunk 7]
```diff
@@ -55,12 +55,15 @@ There are two types of hook handlers that can be implemented for Containers:
 * Exec - Executes a specific command, such as `pre-stop.sh`, inside the cgroups and namespaces of the Container.
 Resources consumed by the command are counted against the Container.
 * HTTP - Executes an HTTP request against a specific endpoint on the Container.
+* Sleep - Executes a sleep command to pause a container for a specified duration.
```
Comment (thockin) [inline content/en/docs/concepts/containers/container-lifecycle-hooks.md:58]: Can we avoid "executes" which implies that `sleep` needs to be a binary in the image?

How about "Sleep - Pauses for a specified duration."

Comment (AxeZhan) [reply content/en/docs/concepts/containers/container-lifecycle-hooks.md]: Updated to `Sleep - Pauses the container for a specified duration. `

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by drewhagen** [comment]
Hello @AxeZhan @thockin :wave:, I see there is work in progress with these docs! 👍 Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday 14th November 2023. Thank you!

**Comment by AxeZhan** [comment]
> Hello @AxeZhan @thockin :wave:, I see there is work in progress with these docs! 👍 Please take a look at [Documenting for a release - PR Ready for Review](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) to get your PR ready for review before Tuesday 14th November 2023. Thank you!


This pr is already merged, so why is this wip? I'm sorry but I'm confused of what to do next.

**Comment by drewhagen** [comment]
@AxeZhan never mind, disregard. Thanks!

**Pull request merged**